In [327]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import re
from scipy import stats

import warnings
warnings.filterwarnings("ignore")
warnings.warn("this will not show")

plt.rcParams["figure.figsize"] = (10, 6)

sns.set_style("whitegrid")
pd.set_option('display.float_format', lambda x: '%.3f' % x)

# Set it None to display all rows in the dataframe
# pd.set_option('display.max_rows', None)

# Set it to None to display all columns in the dataframe
pd.set_option('display.max_columns', None)

In [328]:
df_origin = pd.read_json("scout_car.json",lines=True)
df = df_origin.copy()
df

url      make_model  \
0      https://www.autoscout24.com//offers/audi-a1-sp...         Audi A1   
1      https://www.autoscout24.com//offers/audi-a1-1-...         Audi A1   
2      https://www.autoscout24.com//offers/audi-a1-sp...         Audi A1   
3      https://www.autoscout24.com//offers/audi-a1-1-...         Audi A1   
4      https://www.autoscout24.com//offers/audi-a1-sp...         Audi A1   
...                                                  ...             ...   
15914  https://www.autoscout24.com//offers/renault-es...  Renault Espace   
15915  https://www.autoscout24.com//offers/renault-es...  Renault Espace   
15916  https://www.autoscout24.com//offers/renault-es...  Renault Espace   
15917  https://www.autoscout24.com//offers/renault-es...  Renault Espace   
15918  https://www.autoscout24.com//offers/renault-es...  Renault Espace   

                                       short_description body_type  price  \
0            Sportback 1.4 TDI S-tronic Xenon Navi Klima    Sedans  15770   
1                                         1.8 TFSI sport    Sedans  14500   
2      Sportback 1.6 TDI S tronic Einparkhilfe plus+m...    Sedans  14640   
3                                1.4 TDi Design S tronic    Sedans  14500   
4      Sportback 1.4 TDI S-Tronic S-Line Ext. admired...    Sedans  16790   
...                                                  ...       ...    ...   
15914              Blue dCi 200CV EDC Executive 4Control       Van  39950   
15915                TCe 225 EDC GPF LIM Deluxe Pano,RFK       Van  39885   
15916  Blue dCi 200 EDC Initiale Paris Leder LED Navi...       Van  39875   
15917  Blue dCi 200CV EDC Business , NUOVA DA IMMATRI...       Van  39700   
15918        INITIALE PARIS TCe 225 EDC GPF NAVI SHZ AUX       Van  40999   

                    vat         km registration         prev_owner  kW  \
0        VAT deductible  56,013 km      01/2016  2 previous owners NaN   
1      Price negotiable  80,000 km      03/2017               None NaN   
2        VAT deductible  83,450 km      02/2016   1 previous owner NaN   
3                  None  73,000 km      08/2016   1 previous owner NaN   
4                  None  16,200 km      05/2016   1 previous owner NaN   
...                 ...        ...          ...                ...  ..   
15914    VAT deductible       - km          -/-               None NaN   
15915    VAT deductible   9,900 km      01/2019   1 previous owner NaN   
15916    VAT deductible      15 km      03/2019   1 previous owner NaN   
15917    VAT deductible      10 km      06/2019               None NaN   
15918    VAT deductible       - km      01/2019   1 previous owner NaN   

           hp                                               Type  \
0       66 kW            [, Used, , Diesel (Particulate Filter)]   
1      141 kW                               [, Used, , Gasoline]   
2       85 kW            [, Used, , Diesel (Particulate Filter)]   
3       66 kW            [, Used, , Diesel (Particulate Filter)]   
4       66 kW            [, Used, , Diesel (Particulate Filter)]   
...       ...                                                ...   
15914  147 kW             [, New, , Diesel (Particulate Filter)]   
15915  165 kW  [, Used, , Super 95 / Super Plus 98 (Particula...   
15916  146 kW                       [, Pre-registered, , Diesel]   
15917  147 kW                       [, Pre-registered, , Diesel]   
15918  165 kW                      [, Demonstration, , Super 95]   

                                         Previous Owners  \
0                                                  \n2\n   
1                                                    NaN   
2                                                  \n1\n   
3                                                  \n1\n   
4                                                  \n1\n   
...                                                  ...   
15914                                                NaN   
15915  [\n1\n, \n, 7.4 l/100 

In [329]:
df.sample(5).T

12697  \
url                            https://www.autoscout24.com//offers/opel-insig...   
make_model                                                         Opel Insignia   
short_description                    Grand Sport 1.5 Turbo Business Executive AT   
body_type                                                                 Sedans   
price                                                                      26125   
vat                                                               VAT deductible   
km                                                                          - km   
registration                                                                 -/-   
prev_owner                                                                  None   
kW                                                                           NaN   
hp                                                                        121 kW   
Type                                                     [, New, , Super E10 95]   
Previous Owners                                                              NaN   
Next Inspection                                                              NaN   
Inspection new                                [\nYes\n, \n135 g CO2/km (comb)\n]   
Warranty                                                                     NaN   
Full Service                                                                 NaN   
Non-smoking Vehicle                                   [\n, \n, \nEuro 6d-TEMP\n]   
null                                                                          []   
Make                                                                    \nOpel\n   
Model                                                         [\n, Insignia, \n]   
Offer Number                                                         [\n71263\n]   
First Registration                                                           NaN   
Body Color                                                                   NaN   
Paint Type                                                                   NaN   
Body Color Original                            [\nwählbar - ggf. mit Aufpreis\n]   
Upholstery                                                    [\nCloth, Other\n]   
Body                                                            [\n, Sedans, \n]   
Nr. of Doors                                                             [\n4\n]   
Nr. of Seats                                                             [\n5\n]   
Model Code                                                                   NaN   
Gearing Type                                                 [\n, Automatic, \n]   
Displacement                                                      [\n1,490 cc\n]   
Cylinders                                                                [\n4\n]   
Weight                                                            [\n1,472 kg\n]   
Drive chain                                                          [\nfront\n]   
Fuel                                                      [\n, Super E10 95, \n]   
Consumption                    [[5.9 l/100 km (comb)], [7.6 l/100 km (city)],...   
CO2 Emission                                           [\n135 g CO2/km (comb)\n]   
Emission Class                                                [\nEuro 6d-TEMP\n]   
\nComfort & Convenience\n      [Air conditioning, Armrest, Automatic climate ...   
\nEntertainment & Media\n      [Bluetooth, Digital radio, On-board computer, ...   
\nExtras\n                     [Alloy wheels, Catalytic Converter, Touch screen]   
\nSafety & Security\n          [ABS, Adaptive Cruise Control, Central door lo...   
description                    [\n,  , Neues Fahrzeug inkl.19% MWST, Fahrzeug...   
Emission Label                                                   [\n4 (Green)\n]   
Gears                                                                    [\n6\n]   
Country version                            

In [330]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15919 entries, 0 to 15918
Data columns (total 54 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   url                            15919 non-null  object 
 1   make_model                     15919 non-null  object 
 2   short_description              15873 non-null  object 
 3   body_type                      15859 non-null  object 
 4   price                          15919 non-null  int64  
 5   vat                            11406 non-null  object 
 6   km                             15919 non-null  object 
 7   registration                   15919 non-null  object 
 8   prev_owner                     9091 non-null   object 
 9   kW                             0 non-null      float64
 10  hp                             15919 non-null  object 
 11  Type                           15917 non-null  object 
 12  Previous Owners                9279 non-null  

In [331]:
## bazı sütunlar isimlerinde newline karakterler olduğundan dolayı sıkıntılar var

In [332]:
df.columns

Index(['url', 'make_model', 'short_description', 'body_type', 'price', 'vat',
       'km', 'registration', 'prev_owner', 'kW', 'hp', 'Type',
       'Previous Owners', 'Next Inspection', 'Inspection new', 'Warranty',
       'Full Service', 'Non-smoking Vehicle', 'null', 'Make', 'Model',
       'Offer Number', 'First Registration', 'Body Color', 'Paint Type',
       'Body Color Original', 'Upholstery', 'Body', 'Nr. of Doors',
       'Nr. of Seats', 'Model Code', 'Gearing Type', 'Displacement',
       'Cylinders', 'Weight', 'Drive chain', 'Fuel', 'Consumption',
       'CO2 Emission', 'Emission Class', '\nComfort & Convenience\n',
       '\nEntertainment & Media\n', '\nExtras\n', '\nSafety & Security\n',
       'description', 'Emission Label', 'Gears', 'Country version',
       'Electricity consumption', 'Last Service Date', 'Other Fuel Types',
       'Availability', 'Last Timing Belt Service Date', 'Available from'],
      dtype='object')

In [333]:
 df.rename(columns={"\nEntertainment & Media\n":"Entertainment & Media",
                    '\nComfort & Convenience\n':"Comfort & Convenience",
                    "\nExtras\n" : "Extras",
                    '\nSafety & Security\n':"Safety & Security"
                   
                   
                   }).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15919 entries, 0 to 15918
Data columns (total 54 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   url                            15919 non-null  object 
 1   make_model                     15919 non-null  object 
 2   short_description              15873 non-null  object 
 3   body_type                      15859 non-null  object 
 4   price                          15919 non-null  int64  
 5   vat                            11406 non-null  object 
 6   km                             15919 non-null  object 
 7   registration                   15919 non-null  object 
 8   prev_owner                     9091 non-null   object 
 9   kW                             0 non-null      float64
 10  hp                             15919 non-null  object 
 11  Type                           15917 non-null  object 
 12  Previous Owners                9279 non-null  

In [334]:
 df.rename(columns={"\nEntertainment & Media\n":"Entertainment & Media",
                    '\nComfort & Convenience\n':"Comfort & Convenience",
                    "\nExtras\n" : "Extras",
                    '\nSafety & Security\n':"Safety & Security"
                   
                   
                   },inplace=True)

In [335]:
df.shape

(15919, 54)

In [336]:
df.isna().sum()        # gerçek NaN'ları ve None'ları verecek garip değerleri değil.

url                                  0
make_model                           0
short_description                   46
body_type                           60
price                                0
vat                               4513
km                                   0
registration                         0
prev_owner                        6828
kW                               15919
hp                                   0
Type                                 2
Previous Owners                   6640
Next Inspection                  12384
Inspection new                   11987
Warranty                          5420
Full Service                      7704
Non-smoking Vehicle               8742
null                                 0
Make                                 0
Model                                0
Offer Number                      3175
First Registration                1597
Body Color                         597
Paint Type                        5772
Body Color Original      

Electricity consumption          15782
Last Service Date                15353
Other Fuel Types                 15039
Availability                     15284
Last Timing Belt Service Date    15903
Available from                   15647
kW                               15919                   ## bunlar düşülebilir hemen.

In [337]:
df.isna().sum() / len(df) * 100  

url                               0.000
make_model                        0.000
short_description                 0.289
body_type                         0.377
price                             0.000
vat                              28.350
km                                0.000
registration                      0.000
prev_owner                       42.892
kW                              100.000
hp                                0.000
Type                              0.013
Previous Owners                  41.711
Next Inspection                  77.794
Inspection new                   75.300
Warranty                         34.047
Full Service                     48.395
Non-smoking Vehicle              54.916
null                              0.000
Make                              0.000
Model                             0.000
Offer Number                     19.945
First Registration               10.032
Body Color                        3.750
Paint Type                       36.259


In [338]:
## % 85'den fazla missing value olan sütunlar düşülebilir.

In [339]:
df.drop(labels=["Available from","Last Timing Belt Service Date","Availability","Other Fuel Types","Last Service Date",
               "Electricity consumption","kW"],axis=1).shape

(15919, 47)

In [340]:
df.drop(labels=["Available from","Last Timing Belt Service Date","Availability","Other Fuel Types","Last Service Date",
               "Electricity consumption","kW"],axis=1,inplace=True)

In [341]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15919 entries, 0 to 15918
Data columns (total 47 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   url                    15919 non-null  object
 1   make_model             15919 non-null  object
 2   short_description      15873 non-null  object
 3   body_type              15859 non-null  object
 4   price                  15919 non-null  int64 
 5   vat                    11406 non-null  object
 6   km                     15919 non-null  object
 7   registration           15919 non-null  object
 8   prev_owner             9091 non-null   object
 9   hp                     15919 non-null  object
 10  Type                   15917 non-null  object
 11  Previous Owners        9279 non-null   object
 12  Next Inspection        3535 non-null   object
 13  Inspection new         3932 non-null   object
 14  Warranty               10499 non-null  object
 15  Full Service       

In [342]:
# df.nunique()   # TypeError: unhashable type: 'list  bundan dolayı  str methodlarını daha iyi uygulayabilmem için
                 # listedeki verileri stringe dönüştürmeliyim.

In [343]:
df = df.applymap(lambda x: ','.join(map(str, x)) if isinstance(x, list) else x)
df.sample(2)

url make_model  \
949  https://www.autoscout24.com//offers/audi-a1-sp...    Audi A1   
248  https://www.autoscout24.com//offers/audi-a1-sp...    Audi A1   

                                 short_description body_type  price   vat  \
949           Sportback 1.4 TFSi Pano, Navi, Xenon    Sedans  21400  None   
248  Sportback sport DriveSel°Xenon+°S-Tronic°1.Ha   Compact  16999  None   

            km registration        prev_owner     hp  \
949  15,201 km      05/2017  1 previous owner  92 kW   
248  31,987 km      06/2017  1 previous owner  85 kW   

                                   Type Previous Owners  \
949           ,Used,,Regular/Benzine 91           \n1\n   
248  ,Used,,Diesel (Particulate Filter)           \n1\n   

                         Next Inspection Inspection new          Warranty  \
949  \n05/2020\n,\n123 g CO2/km (comb)\n            NaN  \n,\n,\nEuro 6\n   
248  \n06/2020\n,\n106 g CO2/km (comb)\n            NaN               NaN   

            Full Service  Non-smoking Vehicle null      Make     Model  \
949  \n,\n,\n4 (Green)\n                \n,\n       \nAudi\n  \n,A1,\n   
248     \n,\n,\nEuro 6\n  \n,\n,\n4 (Green)\n       \nAudi\n  \n,A1,\n   

    Offer Number First Registration   Body Color    Paint Type  \
949    \n27547\n         \n,2017,\n  \n,Black,\n  \nMetallic\n   
248          NaN         \n,2017,\n   \n,Blue,\n  \nMetallic\n   

            Body Color Original         Upholstery           Body  \
949  \nMythosschwarz Metallic\n  \nCloth, Orange\n   \n,Sedans,\n   
248     \nUtopiablau Metallic\n   \nCloth, Black\n  \n,Compact,\n   

    Nr. of Doors Nr. of Seats    Model Code     Gearing Type  Displacement  \
949        \n5\n        \n4\n  \n0588/BDC\n  \n,Automatic,\n  \n1,395 cc\n   
248        \n4\n        \n4\n  \n0588/BDG\n  \n,Automatic,\n  \n1,598 cc\n   

    Cylinders Weight Drive chain                               Fuel  \
949       NaN    NaN   \nfront\n           \n,Regular/Benzine 91,\n   
248       NaN    NaN         NaN  \n,Diesel (Particulate Filter),\n   

                                           Consumption  \
949  ['5.2 l/100 km (comb)'],['6.7 l/100 km (city)'...   
248  ['4 l/100 km (comb)'],['4.6 l/100 km (city)'],...   

                CO2 Emission Emission Class  \
949  \n123 g CO2/km (comb)\n     \nEuro 6\n   
248  \n106 g CO2/km (comb)\n     \nEuro 6\n   

                                 Comfort & Convenience  \
949  Air conditioning,Armrest,Automatic climate con...   
248  Air conditioning,Armrest,Automatic climate con...   

                                 Entertainment & Media  \
949  Bluetooth,CD player,Hands-free equipment,On-bo...   
248  Bluetooth,CD player,Hands-free equipment,On-bo...   

                                                Extras  \
949  Alloy wheels,Catalytic Converter,Sport seats,V...   
248  Sport package,Sport seats,Sport suspension,Voi...   

                                     Safety & Security  \
949  ABS,Central door lock,Daytime running lights,D...   
248  ABS,Alarm system,Central door lock,Daytime run...   

                                           description Emission Label Gears  \
949  \n,Highlights:,Anschlussgarantie 1 Jahr, max. ...            NaN   NaN   
248  \nAUDI A1 SPORTBACK SPORT in Utopiablau met.,1...            NaN   NaN   

    Country version  
949     \nGermany\n  
248     \nGermany\n

In [344]:
## bazı \n'leri df'den temizleyebiliriz kolayca : 

In [345]:
df= df.apply(lambda x: x.str.replace("\n","").str.strip(",") if x.dtype == "O" else x)
df.sample(4)

url  make_model  \
4921  https://www.autoscout24.com//offers/audi-a3-sp...     Audi A3   
1808  https://www.autoscout24.com//offers/audi-a1-sp...     Audi A1   
6665  https://www.autoscout24.com//offers/opel-astra...  Opel Astra   
5483  https://www.autoscout24.com//offers/audi-a3-sp...     Audi A3   

                                  short_description body_type  price  \
4921       SPB 30 1.0 TFSI 116CV Sport , NAVI - KM0    Sedans  20900   
1808  Sportback 1.0 TFSI Navi, GRA, Einnparkh. hin.    Sedans  16880   
6665         1.4 Turbo Dynamic INTELLILINK NAVI EU6    Sedans  17440   
5483     Sportback 35 TFSI 150 Nav PDC SHZ Klimaaut    Sedans  23988   

                 vat         km registration        prev_owner      hp  \
4921  VAT deductible      10 km      10/2018              None   85 kW   
1808  VAT deductible  27,200 km      03/2018  1 previous owner   70 kW   
6665  VAT deductible  11,287 km      06/2018              None  110 kW   
5483  VAT deductible      10 km          -/-              None  110 kW   

                                               Type Previous Owners  \
4921  Pre-registered,,Gasoline (Particulate Filter)             NaN   
1808                                 Used,,Super 95               1   
6665                                 Used,,Super 95             NaN   
5483   New,,Regular/Benzine 91 (Particulate Filter)             NaN   

     Next Inspection           Inspection new             Warranty  \
4921             NaN                      NaN                        
1808             NaN  Yes,104 g CO2/km (comb)     36 months,Euro 6   
6665             NaN                      Yes  127 g CO2/km (comb)   
5483             NaN                      Yes  125 g CO2/km (comb)   

      Full Service Non-smoking Vehicle null  Make  Model Offer Number  \
4921           NaN                 NaN       Audi     A3      2056951   
1808     4 (Green)                           Audi     A1        75538   
6665           NaN                 NaN       Opel  Astra  G111918-481   
5483  Euro 6d-TEMP                 NaN       Audi     A3     GW-AZ982   

     First Registration Body Color Paint Type             Body Color Original  \
4921               2018      White        NaN                          Bianco   
1808               2018      Black   Metallic  Mythosschwarz Metallic/Florett   
6665               2018      Brown   Metallic          rouge braun (metallic)   
5483                NaN        Red   Metallic               Tangorot Metallic   

        Upholstery    Body Nr. of Doors Nr. of Seats Model Code Gearing Type  \
4921         Cloth  Sedans            5            5        NaN       Manual   
1808  Cloth, Black  Sedans            5            4   0588/BDB       Manual   
6665  Cloth, Black  Sedans            5            5        NaN    Automatic   
5483         Cloth  Sedans            5            5   0588/BLL       Manual   

     Displacement Cylinders    Weight Drive chain  \
4921       999 cc         3  1,255 kg       front   
1808       999 cc         3  1,135 kg       front   
6665     1,399 cc         4       NaN       front   
5483     1,498 cc       NaN  1,325 kg       front   

                                         Fuel  \
4921            Gasoline (Particulate Filter)   
1808                                 Super 95   
6665                                 Super 95   
5483  Regular/Benzine 91 (Particulate Filter)   

                                            Consumption         CO2 Emission  \
4921  ['5.1 l/100 km (comb)'],['6 l/100 km (city)'],...  117 g CO2/km (comb)   
1808  ['4.4 l/100 km (comb)'],['5.2 l/100 km (city)'...  104 g CO2/km (comb)   
6665  ['5.5 l/100 km (comb)'],['7.2 l/100 km (city)'...  127 g CO2/km (comb)   
5483  ['5.5 l/100 km (comb)'],['6.9 l/100 km (city)'...  125 g CO2/km (comb)   

     Emission Class                              Comfort & Convenience  \
4921         Euro 6  Air conditioning,Automatic climate control,Cru...   
1808         Euro 6  Ai

In [346]:
df.nunique()

url                      15919
make_model                   9
short_description        10001
body_type                    9
price                     2956
vat                          2
km                        6690
registration                48
prev_owner                   4
hp                          81
Type                       169
Previous Owners            101
Next Inspection           1384
Inspection new             201
Warranty                   506
Full Service               121
Non-smoking Vehicle         93
null                         1
Make                         3
Model                        9
Offer Number             11440
First Registration           4
Body Color                  14
Paint Type                   3
Body Color Original       1927
Upholstery                  46
Body                         9
Nr. of Doors                 6
Nr. of Seats                 6
Model Code                 232
Gearing Type                 3
Displacement                77
Cylinder

In [347]:
# df.isin("").sum()

In [348]:
df.isin([""]).sum()   # hangi sütunlarda kaçar tane boş satır var yani bu satırlar da hiçbir şey yok : 

url                          0
make_model                   0
short_description            0
body_type                    0
price                        0
vat                          0
km                           0
registration                 0
prev_owner                   0
hp                           0
Type                         0
Previous Owners              0
Next Inspection              0
Inspection new               0
Warranty                  1343
Full Service              1985
Non-smoking Vehicle       3717
null                     15919
Make                         0
Model                        0
Offer Number                 0
First Registration           0
Body Color                   0
Paint Type                   0
Body Color Original          0
Upholstery                   0
Body                         0
Nr. of Doors                 0
Nr. of Seats                 0
Model Code                   0
Gearing Type                 0
Displacement                 0
Cylinder

In [349]:
## aslında bu bizim için NaN demek. O yüzden ; 

In [350]:
df[df.isin([""])] = np.nan

### 0-) url

In [351]:
df["url"].value_counts()

https://www.autoscout24.com//offers/audi-a1-sportback-1-4-tdi-s-tronic-xenon-navi-klima-diesel-black-bdab349a-caa5-41b0-98eb-c1345b84445e                  1
https://www.autoscout24.com//offers/opel-insignia-bus-inno-2-0-cdti-aut-navi-bi-xe-led-e6-diesel-black-8710b253-dd98-4ed0-9d46-cbad2d79532b                1
https://www.autoscout24.com//offers/opel-insignia-sport-tourer-opel-sport-tourer-1-6-cdti-136-ch-a-c-diesel-silver-10b7463b-9f56-4348-af35-df3432b2b585    1
https://www.autoscout24.com//offers/opel-insignia-2-0cdti-st-business-innovation-navi-bt-diesel-black-ba5a3368-dedc-4e51-9127-3976b26e93c2                 1
https://www.autoscout24.com//offers/opel-insignia-sport-tourer-opel-sport-tourer-1-6-cdti-136-ch-a-c-diesel-silver-7b52f054-78d3-4d5e-9370-8a7d081d7edd    1
                                                                                                                                                          ..
https://www.autoscout24.com//offers/audi-a3-sb-30-1-6-tdi-

In [352]:
## bunun bize yararlı olmayacağını düşünüyorum.

In [353]:
df.drop("url",axis=1,inplace=True)

### 1-) make_model

In [354]:
df["make_model"].value_counts(dropna=False)  
## tam ve düzgün bir feature  gibi belki marka / model diye 2 farklı sütun yapılabilir.

Audi A3           3097
Audi A1           2614
Opel Insignia     2598
Opel Astra        2526
Opel Corsa        2219
Renault Clio      1839
Renault Espace     991
Renault Duster      34
Audi A2              1
Name: make_model, dtype: int64

### 18-) Make

In [355]:
df["Make"].value_counts(dropna=False)

Opel       7343
Audi       5712
Renault    2864
Name: Make, dtype: int64

In [356]:
df["Make"].value_counts(normalize=True)

Opel      0.461
Audi      0.359
Renault   0.180
Name: Make, dtype: float64

### 19-)  Model 

In [357]:
df["Model"]

0            A1
1            A1
2            A1
3            A1
4            A1
          ...  
15914    Espace
15915    Espace
15916    Espace
15917    Espace
15918    Espace
Name: Model, Length: 15919, dtype: object

In [358]:
df["Model"].value_counts()

A3          3097
A1          2614
Insignia    2598
Astra       2526
Corsa       2219
Clio        1839
Espace       991
Duster        34
A2             1
Name: Model, dtype: int64

In [359]:
df.drop(["Model","Make"],axis=1,inplace=True)  

In [360]:
df.shape

(15919, 44)

### 2-)  short_description

In [361]:
df.short_description.value_counts(dropna=False) 

SPB 1.6 TDI 116 CV S tronic Sport                 64
NaN                                               46
1.4 66kW (90CV) Selective                         40
MOVE KLIMA CD USB ALLWETTER BLUETOOTH             38
SPB 30 TDI S tronic Business                      35
                                                  ..
K Sports Tourer Business **Navi*Voll**             1
1.0 Turbo ecoFLEX Automatic Elective Park Navi     1
1.6 CDTI Business *Navi*Automatik*Spurh.*          1
1.6 136 CV Automatica Navi E 6 Garanzia Full       1
TCe 225 EDC GPF LIM Deluxe Pano,RFK                1
Name: short_description, Length: 10002, dtype: int64

In [362]:
## arabalarla ilgili kısa bilgiler içeriyor belki ileride yardımcı olabilir missing_value konusunda o yüzden şimdilik düşmüyorum

### 43-)description    

In [363]:
df["description"].astype("str").value_counts(dropna=False)

nan                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     

In [364]:
## short description gibi bu da belki lazım olabilir.şimdilik tutuyorum.

###  3-) body_type 

In [365]:
df["body_type"].value_counts(dropna=False)    # 60 tane orijinal NaN değerimiz var,sağında solunda temizlenecek bir şey yok
                                              # 290 tane other var.

Sedans           7903
Station wagon    3553
Compact          3153
Van               783
Other             290
Transporter        88
NaN                60
Off-Road           56
Coupe              25
Convertible         8
Name: body_type, dtype: int64

### 26-)Body

In [366]:
df["Body"].value_counts(dropna=False)

Sedans           7903
Station wagon    3553
Compact          3153
Van               783
Other             290
Transporter        88
NaN                60
Off-Road           56
Coupe              25
Convertible         8
Name: Body, dtype: int64

In [367]:
## body_type ile aynı sütunlara ve değerlere sahip olduğundan dolayı bunu bırakacağız.

In [368]:
df.drop(["Body"],axis=1,inplace=True)

In [369]:
df.shape

(15919, 43)

### 5-) vat

VAT deductible : geri talep edilebilir KDV
Price negotiable : Pazarlık edilebilir fiyat

In [370]:
df["vat"].value_counts(dropna=False)

VAT deductible      10980
NaN                  4513
Price negotiable      426
Name: vat, dtype: int64

In [371]:
df["vat"].value_counts(normalize=True,dropna=False)

VAT deductible     0.690
NaN                0.283
Price negotiable   0.027
Name: vat, dtype: float64

In [372]:
## bu sütunda extra temizlenecek bir şey yok ve 4513 tane NaN var.

### 6-) km    

In [373]:
df["km"].value_counts(dropna=False)

10 km        1045
- km         1024
1 km          367
5 km          170
50 km         148
             ... 
67,469 km       1
43,197 km       1
10,027 km       1
35,882 km       1
57 km           1
Name: km, Length: 6690, dtype: int64

In [374]:
df["km"].sample(5)

8444     54,251 km
4371     54,740 km
124      31,750 km
4117     17,412 km
12888         - km
Name: km, dtype: object

In [375]:
# df["km"].str.replace("km","").str.replace(",","").astype("float")

In [376]:
df["km"].str.replace("km","").str.replace(",","").replace("- ",np.nan).astype("float")

0       56013.000
1       80000.000
2       83450.000
3       73000.000
4       16200.000
           ...   
15914         NaN
15915    9900.000
15916      15.000
15917      10.000
15918         NaN
Name: km, Length: 15919, dtype: float64

In [377]:
df["km"] = df["km"].str.replace("km","").str.replace(",","").replace("- ",np.nan).astype("float")

In [378]:
df.km.value_counts(dropna=False).sort_values(ascending=False)

10.000       1045
NaN          1024
1.000         367
5.000         170
50.000        148
             ... 
36980.000       1
36818.000       1
46398.000       1
59535.000       1
57.000          1
Name: km, Length: 6690, dtype: int64

In [379]:
df["km"]

0       56013.000
1       80000.000
2       83450.000
3       73000.000
4       16200.000
           ...   
15914         NaN
15915    9900.000
15916      15.000
15917      10.000
15918         NaN
Name: km, Length: 15919, dtype: float64

In [380]:
## NaN  sayısı 1024  

### 7-) registration : ruhsatın verildiği tarih 

In [381]:
df["registration"].value_counts(dropna=False)

-/-        1597
03/2018     695
02/2019     585
05/2018     572
03/2019     543
01/2019     541
04/2018     541
02/2018     539
03/2016     536
04/2016     532
06/2018     532
01/2018     511
04/2019     506
02/2016     472
03/2017     471
05/2016     459
06/2016     452
05/2019     440
06/2017     409
05/2017     404
07/2018     396
04/2017     380
01/2016     376
02/2017     368
01/2017     306
08/2018     285
06/2019     224
07/2017     215
11/2017     180
07/2016     176
10/2016     160
10/2017     154
09/2017     149
11/2016     142
09/2018     141
09/2016     141
12/2016     134
12/2017     123
08/2017     114
11/2018     110
12/2018     103
10/2018      97
08/2016      94
07/2019       6
09/2019       5
08/2019       1
11/2019       1
12/2019       1
Name: registration, dtype: int64

### 21-) First Registration

In [382]:
df["First Registration"].value_counts(dropna=False)

2018    4522
2016    3674
2017    3273
2019    2853
NaN     1597
Name: First Registration, dtype: int64

In [383]:
# bu iki sütun ruhsat ile ilgili olduğundan arabanın yaşına gidebiliriz buradan ve ayın fazla bir önemi olmadığından ve 
#  registration sütununda da 1597 tane NaN değer olduğundan bu sütunu drop edeceğim.

In [384]:
df.drop("registration",axis=1,inplace=True) 

In [385]:
df.shape

(15919, 42)

### new_feature (age)

In [391]:
# 2019-df["First Registration"]   # assumption . böyle yaparsak hata alırız önce sütunumuzun data type'nı değiştirmeliyiz
                                  # demek ki first registration sütunumuz listelerden oluşuyormuş.

In [392]:
df["First Registration"] =df["First Registration"].astype(float)

In [393]:
df["age"] = 2019-df["First Registration"]

In [394]:
df["age"].value_counts(dropna=False)

1.000    4522
3.000    3674
2.000    3273
0.000    2853
NaN      1597
Name: age, dtype: int64

In [395]:
df.drop("First Registration", axis=1, inplace=True)  # buna da  artık ihtiyacımız kalmadı.

In [396]:
df.shape

(15919, 42)

In [132]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15919 entries, 0 to 15918
Data columns (total 43 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   short_description      15873 non-null  object 
 1   body_type              15859 non-null  object 
 2   price                  15919 non-null  int64  
 3   vat                    11406 non-null  object 
 4   km                     14895 non-null  float64
 5   prev_owner             9091 non-null   object 
 6   hp                     15919 non-null  object 
 7   Type                   15917 non-null  object 
 8   Previous Owners        9279 non-null   object 
 9   Next Inspection        3535 non-null   object 
 10  Inspection new         3932 non-null   object 
 11  Warranty               10499 non-null  object 
 12  Full Service           8215 non-null   object 
 13  Non-smoking Vehicle    7177 non-null   object 
 14  null                   15919 non-null  object 
 15  Ma

### 5- ) prev_owner

In [397]:
df["prev_owner"].value_counts(dropna=False)

1 previous owner     8294
NaN                  6828
2 previous owners     778
3 previous owners      17
4 previous owners       2
Name: prev_owner, dtype: int64

In [398]:
df["prev_owner"].value_counts(dropna=False,normalize=True)

1 previous owner    0.521
NaN                 0.429
2 previous owners   0.049
3 previous owners   0.001
4 previous owners   0.000
Name: prev_owner, dtype: float64

In [403]:
df["prev_owner"].str.extract("(\d)").astype(float)

0
0     2.000
1       NaN
2     1.000
3     1.000
4     1.000
...     ...
15914   NaN
15915 1.000
15916 1.000
15917   NaN
15918 1.000

[15919 rows x 1 columns]

In [404]:
df["prev_owner"] = df["prev_owner"].str.extract("(\d)").astype(float)

###  9-) Previous Owners

In [405]:
df["Previous Owners"].value_counts(dropna=False)

1                                                                      8103
NaN                                                                    6640
2                                                                       766
0                                                                       163
3                                                                        17
                                                                       ... 
1,111 g CO2/km (comb)                                                     1
1,Euro 6                                                                  1
2,119 g CO2/km (comb)                                                     1
1,,6.6 l/100 km (comb),,8.1 l/100 km (city),,5.7 l/100 km (country)       1
1,142 g CO2/km (comb)                                                     1
Name: Previous Owners, Length: 102, dtype: int64

In [406]:
df["Previous Owners"].value_counts(dropna=False,normalize=True)

1                                                                     0.509
NaN                                                                   0.417
2                                                                     0.048
0                                                                     0.010
3                                                                     0.001
                                                                       ... 
1,111 g CO2/km (comb)                                                 0.000
1,Euro 6                                                              0.000
2,119 g CO2/km (comb)                                                 0.000
1,,6.6 l/100 km (comb),,8.1 l/100 km (city),,5.7 l/100 km (country)   0.000
1,142 g CO2/km (comb)                                                 0.000
Name: Previous Owners, Length: 102, dtype: float64

burada consumption (tüketim) ile ilgili bilgiler var doldurma işlemleri sırasında kullanılabilir o yüzden şimdilik drop etmeyelim

###  7-) hp

In [411]:
pd.options.display.max_rows = 100
pd.options.display.max_columns = 100

In [412]:
df["hp"].value_counts(dropna=False)

85 kW     2542
66 kW     2122
81 kW     1402
100 kW    1308
110 kW    1112
70 kW      888
125 kW     707
51 kW      695
55 kW      569
118 kW     516
92 kW      466
121 kW     392
147 kW     380
77 kW      345
56 kW      286
54 kW      276
103 kW     253
87 kW      232
165 kW     194
88 kW      177
60 kW      160
162 kW      98
- kW        88
74 kW       81
96 kW       72
71 kW       59
101 kW      47
67 kW       40
154 kW      39
122 kW      35
119 kW      30
164 kW      27
135 kW      24
82 kW       22
52 kW       22
78 kW       20
1 kW        20
294 kW      18
146 kW      18
141 kW      16
57 kW       10
104 kW       8
120 kW       8
191 kW       7
112 kW       7
155 kW       6
117 kW       6
184 kW       5
90 kW        4
76 kW        4
65 kW        4
149 kW       3
80 kW        3
168 kW       3
98 kW        3
93 kW        3
228 kW       2
270 kW       2
53 kW        2
140 kW       2
86 kW        2
167 kW       2
127 kW       2
89 kW        2
143 kW       2
63 kW        2
40 kW     

In [413]:
df["hp"].value_counts(dropna=False,normalize=True)

85 kW    0.160
66 kW    0.133
81 kW    0.088
100 kW   0.082
110 kW   0.070
70 kW    0.056
125 kW   0.044
51 kW    0.044
55 kW    0.036
118 kW   0.032
92 kW    0.029
121 kW   0.025
147 kW   0.024
77 kW    0.022
56 kW    0.018
54 kW    0.017
103 kW   0.016
87 kW    0.015
165 kW   0.012
88 kW    0.011
60 kW    0.010
162 kW   0.006
- kW     0.006
74 kW    0.005
96 kW    0.005
71 kW    0.004
101 kW   0.003
67 kW    0.003
154 kW   0.002
122 kW   0.002
119 kW   0.002
164 kW   0.002
135 kW   0.002
82 kW    0.001
52 kW    0.001
78 kW    0.001
1 kW     0.001
294 kW   0.001
146 kW   0.001
141 kW   0.001
57 kW    0.001
104 kW   0.001
120 kW   0.001
191 kW   0.000
112 kW   0.000
155 kW   0.000
117 kW   0.000
184 kW   0.000
90 kW    0.000
76 kW    0.000
65 kW    0.000
149 kW   0.000
80 kW    0.000
168 kW   0.000
98 kW    0.000
93 kW    0.000
228 kW   0.000
270 kW   0.000
53 kW    0.000
140 kW   0.000
86 kW    0.000
167 kW   0.000
127 kW   0.000
89 kW    0.000
143 kW   0.000
63 kW    0.000
40 kW    0

In [418]:
df["hp"].str.strip("kW").replace("- ",np.nan).astype(float)

0        66.000
1       141.000
2        85.000
3        66.000
4        66.000
          ...  
15914   147.000
15915   165.000
15916   146.000
15917   147.000
15918   165.000
Name: hp, Length: 15919, dtype: float64

In [423]:
df["hp_kw"] =  df["hp"].str.strip("kW").replace("- ",np.nan).astype(float)  # birimini belirtmeliyiz.

In [421]:
df.drop(columns="hp",inplace=True)

In [424]:
df.shape

(15919, 42)

In [425]:
## biz ilk başta baktığımızda NaN değer yoktu ama garip değerleri NaN'a çevirdik peki kaç tane ? 

In [426]:
df["hp_kw"].value_counts(dropna=False)

85.000     2542
66.000     2122
81.000     1402
100.000    1308
110.000    1112
70.000      888
125.000     707
51.000      695
55.000      569
118.000     516
92.000      466
121.000     392
147.000     380
77.000      345
56.000      286
54.000      276
103.000     253
87.000      232
165.000     194
88.000      177
60.000      160
162.000      98
NaN          88
74.000       81
96.000       72
71.000       59
101.000      47
67.000       40
154.000      39
122.000      35
119.000      30
164.000      27
135.000      24
82.000       22
52.000       22
78.000       20
1.000        20
294.000      18
146.000      18
141.000      16
57.000       10
104.000       8
120.000       8
191.000       7
112.000       7
155.000       6
117.000       6
184.000       5
90.000        4
76.000        4
65.000        4
149.000       3
80.000        3
168.000       3
98.000        3
93.000        3
228.000       2
270.000       2
53.000        2
140.000       2
86.000        2
167.000       2
127.000 

NaN ----> 88

### 8-) Type 

In [428]:
df["Type"].value_counts(dropna=False)

Used,,Diesel (Particulate Filter)                                                                                 3475
Used,,Diesel                                                                                                      2516
Used,,Gasoline                                                                                                    2367
Used,,Super 95                                                                                                    1818
Pre-registered,,Super 95                                                                                           500
                                                                                                                  ... 
New,,Regular/Benzine 91 / Super 95 / Super Plus 98 / Regular/Benzine E10 91 / Super Plus E10 98 / Super E10 95       1
Used,,Super 95 / Super Plus 98 / Super E10 95 / Super Plus E10 98 (Particulate Filter)                               1
Demonstration,,Super Plus 98                    

In [429]:
df["Type"].value_counts(dropna=False,normalize=True)

Used,,Diesel (Particulate Filter)                                                                                0.218
Used,,Diesel                                                                                                     0.158
Used,,Gasoline                                                                                                   0.149
Used,,Super 95                                                                                                   0.114
Pre-registered,,Super 95                                                                                         0.031
                                                                                                                  ... 
New,,Regular/Benzine 91 / Super 95 / Super Plus 98 / Regular/Benzine E10 91 / Super Plus E10 98 / Super E10 95   0.000
Used,,Super 95 / Super Plus 98 / Super E10 95 / Super Plus E10 98 (Particulate Filter)                           0.000
Demonstration,,Super Plus 98                    

In [430]:
df["Type"].isna().sum()

2

 bir de bizim fuel sütunumuz var ileride bu ikisini karşılaştırıp hangisini kullanacaksak oradaki NaN'ları doldurmak için diğerini kullanabiliriz.

In [433]:
#   ex :
df[["Type","Fuel"]].sample(10)

Type                         Fuel
324    Used,,Diesel (Particulate Filter)  Diesel (Particulate Filter)
2979   Used,,Diesel (Particulate Filter)  Diesel (Particulate Filter)
15581                       Used,,Diesel                       Diesel
8018             Demonstration,,Gasoline                     Gasoline
15791                       Used,,Diesel                       Diesel
1615                      Used,,Gasoline                     Gasoline
12062                       Used,,Diesel                       Diesel
4574   Used,,Diesel (Particulate Filter)  Diesel (Particulate Filter)
14466              Demonstration,,Diesel                       Diesel
15439                       Used,,Diesel                       Diesel

In [437]:
df["Type"].str.split(",")

0                    [Used, , Diesel (Particulate Filter)]
1                                       [Used, , Gasoline]
2                    [Used, , Diesel (Particulate Filter)]
3                    [Used, , Diesel (Particulate Filter)]
4                    [Used, , Diesel (Particulate Filter)]
                               ...                        
15914                 [New, , Diesel (Particulate Filter)]
15915    [Used, , Super 95 / Super Plus 98 (Particulate...
15916                           [Pre-registered, , Diesel]
15917                           [Pre-registered, , Diesel]
15918                          [Demonstration, , Super 95]
Name: Type, Length: 15919, dtype: object

In [440]:
df["Type"].str.split(",").str[0]  # burada mesela liste yeniden işimize yaradı.

0                  Used
1                  Used
2                  Used
3                  Used
4                  Used
              ...      
15914               New
15915              Used
15916    Pre-registered
15917    Pre-registered
15918     Demonstration
Name: Type, Length: 15919, dtype: object

In [441]:
df["Type_"] = df["Type"].str.split(",").str[0]

In [443]:
df["Type_"].value_counts(dropna=False)               

Used              11096
New                1650
Pre-registered     1364
Employee's car     1011
Demonstration       796
NaN                   2
Name: Type_, dtype: int64

NaN ------> 2

In [444]:
df.shape

(15919, 43)

### 10-) Next Inspection   ---> sonraki muayene tarihi

In [460]:
df.rename(columns={"Next Inspection":"next_inspection"},inplace=True)

In [461]:
df["next_inspection"].value_counts(dropna=False)

NaN                            12384
04/2022                           62
06/2021                           47
03/2021                           39
03/2022                           36
                               ...  
06/2021,121 g CO2/km (comb)        1
02/2020,139 g CO2/km (comb)        1
10/2021,136 g CO2/km (comb)        1
06/2019,139 g CO2/km (comb)        1
01/2022,168 g CO2/km (comb)        1
Name: next_inspection, Length: 1385, dtype: int64

In [462]:
df["next_inspection"].value_counts(dropna=False,normalize=True)

NaN                           0.778
04/2022                       0.004
06/2021                       0.003
03/2021                       0.002
03/2022                       0.002
                               ... 
06/2021,121 g CO2/km (comb)   0.000
02/2020,139 g CO2/km (comb)   0.000
10/2021,136 g CO2/km (comb)   0.000
06/2019,139 g CO2/km (comb)   0.000
01/2022,168 g CO2/km (comb)   0.000
Name: next_inspection, Length: 1385, dtype: float64

In [463]:
# NaN DEĞER BAYA FAZLA.

In [472]:
df.next_inspection.str.extract("(\d+/\d{0,4})").isna().sum() 
## tarihler ayrı bir sütun yapılsa 12384 tane null değer var şimdilik böyle bırakıyorum.                                                             

0    12384
dtype: int64

In [464]:
## bir de bizim inspection new diye bir column'umuz var: 

### 11 -) Inspection new 

In [465]:
df.rename(columns={"Inspection new":"inspection_new"},inplace=True)

In [466]:
df["inspection_new"].isna().sum()

11987

In [467]:
df[["next_inspection","inspection_new"]].sample(10)

next_inspection           inspection_new
6000                       05/2020                      NaN
13                             NaN                      NaN
911                            NaN  Yes,0 kWh/100 km (comb)
5742                           NaN  Yes,119 g CO2/km (comb)
7767                           NaN                      NaN
12648                          NaN                      NaN
8314                           NaN                      NaN
11816                          NaN  Yes,141 g CO2/km (comb)
5841   03/2021,124 g CO2/km (comb)                      NaN
1484                           NaN  Yes,102 g CO2/km (comb)

In [468]:
## bir benzerlik bulunamadı aralarında.

### 12-)  Warranty

In [474]:
df["Warranty"].value_counts(dropna=False)

NaN                                                                          6763
Euro 6                                                                       1868
12 months                                                                    1305
24 months                                                                     585
4 (Green)                                                                     517
                                                                             ... 
2 months                                                                        1
72 months,125 g CO2/km (comb)                                                   1
60 months,14 g CO2/km (comb)                                                    1
24 months,121 g CO2/km (comb)                                                   1
36 months,,7.4 l/100 km (comb),,9.6 l/100 km (city),,6 l/100 km (country)       1
Name: Warranty, Length: 506, dtype: int64

In [473]:
df["Warranty"].value_counts(dropna=False,normalize=True)

NaN                                                                         0.425
Euro 6                                                                      0.117
12 months                                                                   0.082
24 months                                                                   0.037
4 (Green)                                                                   0.032
                                                                             ... 
2 months                                                                    0.000
72 months,125 g CO2/km (comb)                                               0.000
60 months,14 g CO2/km (comb)                                                0.000
24 months,121 g CO2/km (comb)                                               0.000
36 months,,7.4 l/100 km (comb),,9.6 l/100 km (city),,6 l/100 km (country)   0.000
Name: Warranty, Length: 506, dtype: float64

In [494]:
df["Warranty"].str.extract(r"(\d+)\s*month").astype(float)

0
0        NaN
1        NaN
2        NaN
3        NaN
4        NaN
...      ...
15914 24.000
15915    NaN
15916    NaN
15917    NaN
15918    NaN

[15919 rows x 1 columns]

In [495]:
df["warranty_month"] = df["Warranty"].str.extract(r"(\d+)\s*month").astype(float)

In [497]:
df["warranty_month"].isnull().sum()      # ÇOK FAZLA NULL DEĞER VAR YİNE.

11066

### 13-) Full Service

In [498]:
df.rename(columns={"Full Service":"full_service"},inplace=1)

In [505]:
df["full_service"].value_counts(dropna=False)

NaN                                                                 9689
4 (Green)                                                           2235
Euro 6                                                              2097
Euro 6d-TEMP                                                         399
120 g CO2/km (comb)                                                  105
                                                                    ... 
6 l/100 km (comb),,8 l/100 km (city),,4.9 l/100 km (country)           1
84 g CO2/km (comb)                                                     1
87 g CO2/km (comb)                                                     1
5.7 l/100 km (comb),,7.3 l/100 km (city),,4.9 l/100 km (country)       1
Euro 4                                                                 1
Name: full_service, Length: 121, dtype: int64

In [500]:
df["full_service"].value_counts(dropna=False,normalize=True)

NaN                                                                0.609
4 (Green)                                                          0.140
Euro 6                                                             0.132
Euro 6d-TEMP                                                       0.025
120 g CO2/km (comb)                                                0.007
                                                                    ... 
6 l/100 km (comb),,8 l/100 km (city),,4.9 l/100 km (country)       0.000
84 g CO2/km (comb)                                                 0.000
87 g CO2/km (comb)                                                 0.000
5.7 l/100 km (comb),,7.3 l/100 km (city),,4.9 l/100 km (country)   0.000
Euro 4                                                             0.000
Name: full_service, Length: 121, dtype: float64

### DİKKAT : SÜTUNLARI İNCELERKEN HEP AYNI ŞEYLERİ YAPTIĞIMIZDAN DOLAYI GENELLEŞTİRME İHİYACI
### AKLIMA GELDİ  BİR SÜTUNU ALDIĞIMIZDA BİR FONKSİYON YARDIMIYLA BAKABİLİRİZ BUNLARA:

In [508]:
def summary(dataframe,col_name):
    print(pd.DataFrame({col_name: dataframe[col_name].value_counts(dropna=False),
           "oran" : (dataframe[col_name].value_counts(normalize=True,dropna=False) * 100)}))
    print("########################################")
    print('Number_of_Nulls   : ', dataframe[col_name].isnull().sum())

In [509]:
summary(df,"full_service")

                                                    full_service   oran
NaN                                                         9689 60.864
4 (Green)                                                   2235 14.040
Euro 6                                                      2097 13.173
Euro 6d-TEMP                                                 399  2.506
120 g CO2/km (comb)                                          105  0.660
...                                                          ...    ...
6 l/100 km (comb),,8 l/100 km (city),,4.9 l/100...             1  0.006
84 g CO2/km (comb)                                             1  0.006
87 g CO2/km (comb)                                             1  0.006
5.7 l/100 km (comb),,7.3 l/100 km (city),,4.9 l...             1  0.006
Euro 4                                                         1  0.006

[121 rows x 2 columns]
########################################
Number_of_Nulls   :  9689


### 14-) Non-smoking Vehicle

In [512]:
df.rename({"Non-smoking Vehicle" : "non_smoking_vehicle"},axis=1,inplace=1)

In [513]:
summary(df,"non_smoking_vehicle")

                                                    non_smoking_vehicle   oran
NaN                                                               12459 78.265
4 (Green)                                                          1240  7.789
Euro 6                                                             1127  7.080
Euro 6d-TEMP                                                        345  2.167
120 g CO2/km (comb)                                                  63  0.396
85 g CO2/km (comb)                                                   46  0.289
107 g CO2/km (comb)                                                  36  0.226
104 g CO2/km (comb)                                                  36  0.226
153 g CO2/km (comb)                                                  31  0.195
124 g CO2/km (comb)                                                  29  0.182
1 (No sticker)                                                       24  0.151
99 g CO2/km (comb)                                  

In [514]:
## null değer 12459 *****  ayrıca içerikle sütun örtüşmüyor ama diğer sütunların doldurulmasında kullanılabilir.

###  15-)  null  

In [516]:
summary(df,"null")

      null    oran
NaN  15919 100.000
########################################
Number_of_Nulls   :  15919


In [519]:
# bunu önceden atmayı unutmuşuz.
df.drop("null",axis=1,inplace=True)

### 18-) Offer Number

In [521]:
summary(df,"Offer Number")

                                          Offer Number   oran
NaN                                               3175 19.945
LT67679                                             27  0.170
UN89904                                             27  0.170
XJ38068                                             27  0.170
JV03654                                             27  0.170
...                                                ...    ...
160_dcbb6c3e-a6da-43a3-8754-ccd994cec93b             1  0.006
6701576                                              1  0.006
6701569                                              1  0.006
158121                                               1  0.006
Re_30000008029                                       1  0.006

[11441 rows x 2 columns]
########################################
Number_of_Nulls   :  3175


In [522]:
## url sütunu gibi bir sütun bizim için o yüzden düşeceğiz : 

In [523]:
df.drop("Offer Number",axis=1,inplace=True)

### 20-) Body Color

In [525]:
df.rename({"Body Color":"body_color"},axis=1,inplace=True)

In [526]:
summary(df,"body_color")

        body_color   oran
Black         3745 23.525
Grey          3505 22.018
White         3406 21.396
Silver        1647 10.346
Blue          1431  8.989
Red            957  6.012
NaN            597  3.750
Brown          289  1.815
Green          154  0.967
Beige          108  0.678
Yellow          51  0.320
Violet          18  0.113
Bronze           6  0.038
Orange           3  0.019
Gold             2  0.013
########################################
Number_of_Nulls   :  597


In [527]:
## ekstra temizlenecek bir şey yok nullar var onlar doldurulabilir mi vs ona ileride bakılacak.

### 22-) Body Color Original

In [528]:
df.rename({"Body Color Original":"body_color_orig"},axis=1,inplace=True)

In [529]:
summary(df,"body_color_orig")

                               body_color_orig   oran
NaN                                       3759 23.613
Onyx Schwarz                               338  2.123
Bianco                                     282  1.771
Mythosschwarz Metallic                     238  1.495
Brillantschwarz                            216  1.357
...                                        ...    ...
Rouge-Braun (G0Y)                            1  0.006
VARI COLRI DISPONIBILI                       1  0.006
Kokosnussbraun Metallic                      1  0.006
Farbe frei wählbar                           1  0.006
Perlmutt-Weiß Metallic (Weiß)                1  0.006

[1928 rows x 2 columns]
########################################
Number_of_Nulls   :  3759


### 21-)  Paint Type

In [530]:
df.rename({"Paint Type":"paint_type"},axis=1,inplace=True)

In [531]:
summary(df,"paint_type")

             paint_type   oran
Metallic           9794 61.524
NaN                5772 36.259
Uni/basic           347  2.180
Perl effect           6  0.038
########################################
Number_of_Nulls   :  5772


In [532]:
## şimdi bu  3 column birbiriyle ilintili olabilir bunlara beraber bakacak olursak : 

In [535]:
df[["paint_type","body_color_orig","body_color"]].sample(15)

paint_type                 body_color_orig body_color
13593        NaN                             NaN      White
11340   Metallic                      Quarz Grau       Grey
2978         NaN                          Blanco      White
14340        NaN                             NaN      White
15542   Metallic                             NaN      Black
1306    Metallic                             NaN       Blue
13152   Metallic                       CASSIOPEE       Grey
3844    Metallic          Florettsilber Metallic     Silver
13782        NaN                             NaN      White
5112    Metallic          Blu Cosmo Metallizzato       Blue
1662    Metallic                             Blu       Blue
14085   Metallic                             NaN       Grey
6507    Metallic                  Blauw Metallic       Blue
2451    Metallic          Manhattangrau Metallic       Grey
11479        NaN  BLACK MEET KETTLE MET-4 (507B)        NaN

### acaba body_color sütunundaki boş değerleri body_color_origin ile doldurabilir miyim ? 

In [547]:
df.loc[df["body_color"].isna(),["body_color","body_color_orig"]]

body_color  body_color_orig
70           NaN        gris nano
122          NaN          wählbar
150          NaN      BLEU UTOPIA
195          NaN      BLEU UTOPIA
198          NaN             null
...          ...              ...
15690        NaN              NaN
15785        NaN  GNE NOIR ETOILE
15849        NaN  Farbe: Sonstige
15854        NaN              NaN
15864        NaN              NaN

[597 rows x 2 columns]

In [548]:
## bunu part2'de kullanabiliriz.

### 23-) Upholstery (döşeme)

In [550]:
summary(df,"Upholstery")

                     Upholstery   oran
Cloth, Black               5821 36.566
NaN                        3720 23.368
Part leather, Black        1121  7.042
Cloth                      1005  6.313
Cloth, Grey                 891  5.597
Cloth, Other                639  4.014
Full leather, Black         575  3.612
Black                       491  3.084
Grey                        273  1.715
Other, Other                182  1.143
Part leather                140  0.879
Full leather                139  0.873
Full leather, Brown         116  0.729
Part leather, Grey          116  0.729
Other, Black                110  0.691
Full leather, Other          72  0.452
Full leather, Grey           67  0.421
Part leather, Other          65  0.408
Other                        56  0.352
Part leather, Brown          50  0.314
alcantara, Black             47  0.295
Velour, Black                36  0.226
Full leather, Beige          36  0.226
Cloth, Brown                 28  0.176
Velour                   

In [561]:
## deri , kumaş , other , kadife gibi döşeme türleri var ve
## bunların renkleri var biz bunları 2 ayrı sütun gibi ele alabiliriz.

In [580]:
df["Upholstery"].str.split(", ", expand=True)

0      1
0             Cloth  Black
1             Cloth   Grey
2             Cloth  Black
3               NaN    NaN
4             Cloth  Black
...             ...    ...
15914           NaN    NaN
15915         Cloth   None
15916  Full leather  Black
15917  Part leather   None
15918  Full leather  Brown

[15919 rows x 2 columns]

In [581]:
df[["upholstery_type","upholstery_color"]] = df["Upholstery"].str.split(", ", expand=True)

In [582]:
summary(df,"upholstery_type")

              upholstery_type   oran
Cloth                    8423 52.912
NaN                      3720 23.368
Part leather             1499  9.416
Full leather             1009  6.338
Black                     491  3.084
Other                     368  2.312
Grey                      273  1.715
Velour                     60  0.377
alcantara                  57  0.358
Brown                      12  0.075
Beige                       3  0.019
Blue                        2  0.013
White                       2  0.013
########################################
Number_of_Nulls   :  3720


In [583]:
summary(df,"upholstery_color")

        upholstery_color   oran
Black               7710 48.433
NaN                 5861 36.818
Grey                1103  6.929
Other                960  6.031
Brown                195  1.225
Beige                 51  0.320
Blue                  14  0.088
White                 11  0.069
Red                    9  0.057
Yellow                 4  0.025
Orange                 1  0.006
########################################
Number_of_Nulls   :  5861


### bu sütuna tekrar BAKALIM

### 24-) Nr. of Doors

In [584]:
df.rename({"Nr. of Doors":"number_of_doors"},axis=1,inplace=True)

In [585]:
summary(df,"number_of_doors")

     number_of_doors   oran
5              11575 72.712
4               3079 19.342
3                832  5.226
2                219  1.376
NaN              212  1.332
1                  1  0.006
7                  1  0.006
########################################
Number_of_Nulls   :  212


In [587]:
df["number_of_doors"]   # dtype'ı değiştirilebilir temizlenecek bir şey yok başka.

0        5
1        3
2        4
3        3
4        5
        ..
15914    5
15915    5
15916    5
15917    5
15918    5
Name: number_of_doors, Length: 15919, dtype: object

In [588]:
df["number_of_doors"]= df["number_of_doors"].astype(float)

### 25 -) Nr. of Seats

In [589]:
df.rename({"Nr. of Seats":"number_of_seats"},axis=1,inplace=True)

In [590]:
summary(df,"number_of_seats")

     number_of_seats   oran
5              13336 83.774
4               1125  7.067
NaN              977  6.137
7                362  2.274
2                116  0.729
6                  2  0.013
3                  1  0.006
########################################
Number_of_Nulls   :  977


In [592]:
df["number_of_seats"]    # bunun da sadece dtype'ını değiştirmekle yetinilebiliriz

0        5
1        4
2        4
3        4
4        5
        ..
15914    5
15915    5
15916    7
15917    7
15918    5
Name: number_of_seats, Length: 15919, dtype: object

In [593]:
df["number_of_seats"]= df["number_of_seats"].astype(float)

### 26-) Model Code

In [594]:
df.rename({"Model Code":"model_code"},axis=1,inplace=True)

In [595]:
summary(df,"model_code")

          model_code   oran
NaN            10941 68.729
0035/BCB         268  1.684
0588/BNO         245  1.539
0588/BDB         206  1.294
0588/BHX         188  1.181
...              ...    ...
0035/AVN           1  0.006
0035/BEM           1  0.006
0035/BAG           1  0.006
0035/BAA           1  0.006
3333/BKG           1  0.006

[233 rows x 2 columns]
########################################
Number_of_Nulls   :  10941


In [596]:
## url sütunu gibi bizim için düşebiliriz.

In [597]:
df.drop(columns="model_code", inplace=True)

In [598]:
df.shape

(15919, 43)

### 27 -) Gearing Type

In [605]:
df.rename({"Gearing Type":"gearing_type"},axis=1,inplace=True)

In [606]:
summary(df,)

0        Automatic
1        Automatic
2        Automatic
3        Automatic
4        Automatic
           ...    
15914    Automatic
15915    Automatic
15916    Automatic
15917    Automatic
15918    Automatic
Name: gearing_type, Length: 15919, dtype: object

In [511]:
## df[df.full_service.notna()][["full_service","co_2_emission","emission_class", "emission_label"]].sample(50)

In [257]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15919 entries, 0 to 15918
Data columns (total 44 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   short_description      15873 non-null  object
 1   body_type              15859 non-null  object
 2   price                  15919 non-null  int64 
 3   vat                    11406 non-null  object
 4   km                     15919 non-null  object
 5   registration           15919 non-null  object
 6   prev_owner             9091 non-null   object
 7   hp                     15919 non-null  object
 8   Type                   15917 non-null  object
 9   Previous Owners        9279 non-null   object
 10  Next Inspection        3535 non-null   object
 11  Inspection new         3932 non-null   object
 12  Warranty               10499 non-null  object
 13  Full Service           8215 non-null   object
 14  Non-smoking Vehicle    7177 non-null   object
 15  null               

### 2-) price ------> target feature